### PR TITLE
Small refactor of Parameter Parsing

### PR DESCRIPTION
**Pull request by pksunkara**
This PR does several things:

1. Moves old style parameter parsing to be done by signature parser.
2. Defaults some parameters to newer style syntax
3. Fixes #123
4. Improves the code. :smile:

### CODE REVIEW

**Comment by pksunkara** [comment]
@w-vi @klokane Please review & merge.

File: src/ParameterParser.h [hunk 1]
```diff
@@ -196,176 +234,108 @@ namespace snowcrash {
...
 
-                mdp::ByteBuffer innerSignature = signature;
-                innerSignature = TrimString(innerSignature);
+            if (attributes.size() <= max) {
+                size_t i = 0;
+                bool definedUse = false;
 
-                size_t firstSpace = innerSignature.find(" ");
+                // Traverse over parameter's traits
+                while (i < attributes.size()) {
```
Comment (kylef) [inline src/ParameterParser.h]: What about using a `for` loop since there is an increment below?

```diff
- while (i < attributes.size()) {
+ for (; i < attributes.size(); ++i) {
...
- i++;
```

Comment (pksunkara) [inline src/ParameterParser.h]: It started out as a `while` since I needed something. But I guess, it can now be `for`.

Comment (w-vi) [inline src/ParameterParser.h]: please rewrite it as for

File: src/ParameterParser.h [hunk 2]
```diff
@@ -196,176 +234,108 @@ namespace snowcrash {
...
 
             out.node.use = UndefinedParameterUse;
-            TrimString(signature);
-
-            if (pd.sectionContext() == ParameterSectionType) {
+            size_t max = oldSyntax ? OLD_SYNTAX_MAX_ATTRIBUTES : NEW_SYNTAX_MAX_ATTRIBUTES;
 
-                mdp::ByteBuffer innerSignature = signature;
-                innerSignature = TrimString(innerSignature);
+            if (attributes.size() <= max) {
```
Comment (klokane) [inline src/ParameterParser.h]: you can swap this condition into:
```
if (attributes.size() > max) {
  warnAboutAdditionalTraits(node, pd, out, oldSyntax);
  return;
}
```
and you will get one level nested less. We will get better readability

File: test/test-ParametersParser.cc [hunk 3]
```diff
@@ -84,7 +84,7 @@ TEST_CASE("Parse illegal parameter among legal ones", "[parameters]")
     "+ Parameters\n"\
     "    + OK_1\n"\
     "    + i;legal\n"\
-    "    + OK-2\n";
+    "    + OK_2\n";
```
Comment (klokane) [inline test/test-ParametersParser.cc]: why is this changed?

**Comment by klokane** [comment]
No other comments from me

File: test/test-ParametersParser.cc (hunk 3 above)
Comment (pksunkara) [inline test/test-ParametersParser.cc]: With this PR, whenever we find ambiguous parameters, we will be defaulting to new syntax assumption.

In the new MSON aligned syntax `OK-2` is `OK` parameter name and `2` description. So, I edited it to `OK_2` so that `OK_2` is parameter name.

**Comment by pksunkara** [comment]
Updated.

File: test/test-ParametersParser.cc (hunk 3 above)
Comment (klokane) [inline test/test-ParametersParser.cc]: do I understand correctly, it changes behavior for old syntax?
If yes, I mean we should add some kind of deprecation warning.

Comment (pksunkara) [inline test/test-ParametersParser.cc]: It changes the behaviour of old syntax only if the parameter has it's name and nothing else and that too there's a `-` in the name. Which is very rare and I think it should still parsed as MSON since `-` should expressed in URI encoded scheme.
